File: rust/crates/kit/src/mpp/protocol/intents/subscription.rs
//! Subscription intent request type and supporting payloads.
//!
//! The subscription intent represents a recurring fixed-amount payment
//! authorized once per billing period through an on-chain delegation.
//! Activation atomically creates the delegation and executes the first
//! billing-period charge; renewals are server-driven on-chain transactions
//! and do not produce HTTP credentials.

use serde::{Deserialize, Serialize};

use crate::mpp::error::Error;

/// Domain separator for reusable subscription authentication proofs.
pub const SUBSCRIPTION_AUTHENTICATION_DOMAIN: &str = "mpp-subscription-auth-v1";

/// Reusable payer proof bound to one activation challenge and delegation.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubscriptionAuthentication {
    /// Always `proof` on the wire.
    #[serde(rename = "type")]
    pub kind: SubscriptionAuthenticationType,
    /// Activation challenge identifier signed into the proof.
    pub challenge_id: String,
    /// Subscriber public key (base58).
    pub payer: String,
    /// Ed25519 signature over the canonical authentication message (base58).
    pub signature: String,
}

/// Discriminator for a reusable subscription proof.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SubscriptionAuthenticationType {
    /// An Ed25519 payer proof.
    Proof,
}

impl SubscriptionAuthentication {
    /// Return the RFC 8785/JCS message bytes signed by the subscriber.
    pub fn message_bytes(&self, subscription_delegation: &str) -> Result<Vec<u8>, Error> {
        let value = serde_json::json!({
            "domain": SUBSCRIPTION_AUTHENTICATION_DOMAIN,
            "payer": self.payer,
            "subscriptionChallengeId": self.challenge_id,
            "subscriptionDelegation": subscription_delegation,
        });
        serde_json_canonicalizer::to_vec(&value).map_err(|error| Error::Other(error.to_string()))
    }

    /// Verify this proof against its payer and bound delegation.
    pub fn verify(&self, subscription_delegation: &str) -> Result<bool, Error> {
        use ed25519_dalek::{Signature, Verifier, VerifyingKey};

        let payer: [u8; 32] = bs58::decode(&self.payer)
            .into_vec()
            .map_err(|error| Error::Other(error.to_string()))?
            .try_into()
            .map_err(|_| Error::Other("payer must be 32 bytes".to_string()))?;
        let signature: [u8; 64] = bs58::decode(&self.signature)
            .into_vec()
            .map_err(|error| Error::Other(error.to_string()))?
            .try_into()
            .map_err(|_| Error::Other("signature must be 64 bytes".to_string()))?;
        let key =
            VerifyingKey::from_bytes(&payer).map_err(|error| Error::Other(error.to_string()))?;
        Ok(key
            .verify(
                &self.message_bytes(subscription_delegation)?,
                &Signature::from_bytes(&signature),
            )
            .is_ok())
    }
}

/// Billing period unit. The Solana profile supports `day` and `week` only;
/// `month` is rejected because the on-chain program uses fixed elapsed
/// seconds and cannot represent calendar-month cadence exactly.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub enum SubscriptionPeriodUnit {
    #[default]
    Day,
    Week,
}

impl SubscriptionPeriodUnit {
    /// Map a `(unit, count)` pair to the subscriptions program's
    /// `period_hours` value.
    ///
    /// Returns an error if the count is out of range for the unit or if the
    /// resulting `period_hours` exceeds the program's `[1, 8760]` bound.
    pub fn to_period_hours(self, period_count: u64) -> Result<u64, Error> {
        if period_count == 0 {
            return Err(Error::Other(
                "periodCount must be a positive integer".into(),
            ));
        }
        match self {
            SubscriptionPeriodUnit::Day => {
                if period_count > 365 {
                    return Err(Error::Other(format!(
                        "periodCount={period_count} for periodUnit=\"day\" exceeds 365"
                    )));
                }
                Ok(period_count * 24)
            }
            SubscriptionPeriodUnit::Week => {
                if period_count > 52 {
                    return Err(Error::Other(format!(
                        "periodCount={period_count} for periodUnit=\"week\" exceeds 52"
                    )));
                }
                Ok(period_count * 168)
            }
        }
    }
}

/// Subscription request (for the `subscription` intent on Solana).
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SubscriptionRequest {
    /// Per-period token amount in base units.
    pub amount: String,

    /// Base58 SPL token mint address (canonical wire form). Implementations
    /// MUST treat this consistently with `methodDetails.mint`.
    pub currency: String,

    /// Billing period unit. The Solana profile supports `day` and `week`
    /// only; `month` is rejected at the schema layer.
    pub period_unit: SubscriptionPeriodUnit,

    /// Decimal string count of `period_unit` values per billing period.
    pub period_count: String,

    /// Primary recipient wallet pubkey (base58).
    pub recipient: String,

    /// Optional RFC3339 expiry of the recurring authorization.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subscription_expires: Option<String>,

    /// Human-readable description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    /// Merchant reference.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub external_id: Option<String>,

    /// Solana-specific extension fields.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub method_details: Option<serde_json::Value>,
}

impl SubscriptionRequest {
    /// Parse `period_count` as a `u64`.
    pub fn parse_period_count(&self) -> Result<u64, Error> {
        self.period_count
            .parse()
            .map_err(|_| Error::Other(format!("Invalid periodCount: {}", self.period_count)))
    }

    /// Parse `amount` as a `u64`.
    pub fn parse_amount(&self) -> Result<u64, Error> {
        self.amount
            .parse()
            .map_err(|_| Error::Other(format!("Invalid amount: {}", self.amount)))
    }

    /// Compute the on-chain `period_hours` for this request, validating both
    /// the period mapping and the program-level bound `[1, 8760]`.
    pub fn period_hours(&self) -> Result<u64, Error> {
        let count = self.parse_period_count()?;
        let hours = self.period_unit.to_period_hours(count)?;
        if hours == 0 || hours > 8760 {
            return Err(Error::Other(format!(
                "period_hours {hours} out of [1, 8760] range"
            )));
        }
        Ok(hours)
    }
}

/// Single-action credential payload. The Solana subscription profile only
/// defines `activate`; renewals are server-driven and cancellations are
/// out-of-band on-chain.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "camelCase")]
pub enum SubscriptionAction {
    /// Activation: subscribe + first-period charge in one transaction.
    Activate(ActivatePayload),
}

/// Activation payload. Mirrors the Solana charge profile's two-mode shape:
/// `type="transaction"` (server broadcasts) or `type="signature"`
/// (client already broadcast).
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ActivatePayload {
    /// Payload type discriminator: `"transaction"` or `"signature"`.
    #[serde(rename = "type")]
    pub payload_type: String,

    /// Standard base64 of the serialized activation transaction
    /// (when `type="transaction"`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transaction: Option<String>,

    /// Base58 of the on-chain transaction signature (when `type="signature"`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signature: Option<String>,

    /// Reusable subscriber proof bound to the activation challenge and
    /// resulting `SubscriptionDelegation` PDA.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub authentication: Option<SubscriptionAuthentication>,
}

/// Credential payload for later access under an active subscription.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubscriptionAccessPayload {
    /// Always `proof` on the wire.
    #[serde(rename = "type")]
    pub payload_type: SubscriptionAuthenticationType,
    /// Base58 `SubscriptionDelegation` PDA bound by the proof.
    pub subscription_delegation: String,
    /// Reusable subscriber proof retained from activation.
    pub authentication: SubscriptionAuthentication,
}

/// Extension fields placed on the standard Receipt's metadata for a
/// subscription activation or renewal.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SubscriptionReceiptExtensions {
    /// Server-issued opaque base64url subscription identifier.
    pub subscription_id: String,
    /// Base58 address of the on-chain SubscriptionDelegation account.
    pub subscription_delegation: String,
    /// Decimal index of the billing period (0 for activation).
    pub period_index: u64,
    /// RFC3339 timestamp of the current period's start.
    pub period_start: String,
    /// RFC3339 timestamp of the current period's end (exclusive).
    pub period_end: String,
    /// RFC3339 effective subscription expiry, when set.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,
}

/// Typed `methodDetails` payload for the Solana subscription intent.
///
/// Serialised in camelCase per the spec wire form. The server (which
/// builds + HMAC-pins it into the 402 challenge) and the client (which
/// reads it back to construct the activation transaction) both use this
/// struct directly — no manual JSON pokery.
///
/// The `expected_*` and `plan_*` server-extension fields mirror the
/// immutable Plan terms the on-chain `Subscribe` instruction needs in
/// its `SubscribeData` payload. Including them in the challenge lets the
/// client build a settle-able activation transaction without an extra
/// RPC roundtrip to fetch the Plan account. When absent the client falls
/// back to RPC.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubscriptionMethodDetails {
    /// Base58 address of the on-chain `Plan` account.
    pub plan_address: String,
    /// Base58 of the SPL token mint. MUST equal the on-chain `plan.mint`.
    pub mint: String,
    /// Base58 of the SPL Token / Token-2022 program id used for the
    /// per-period transfers.
    pub token_program: String,
    /// Decimal precision of the mint. Server-populated; clients use it
    /// to render the amount in human form.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub decimals: Option<u8>,
    /// Base58 of the server's puller pubkey. MUST be `plan.owner` or
    /// appear in `plan.pullers`.
    pub puller: String,
    /// Base58 of the Plan's owner (the merchant who published it). The
    /// on-chain `Subscribe` instruction needs this as its second
    /// account meta and as part of the Plan PDA derivation. When unset,
    /// the client falls back to `puller` (correct only when the
    /// merchant published the plan AND is its own puller).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub merchant: Option<String>,
    /// Base58 of the recipient wallet — must be in `plan.destinations`
    /// (or the whitelist must be empty). The transfer_subscription
    /// instruction routes the first-period charge to this wallet's
    /// associated token account.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recipient: Option<String>,
    /// The per-period charge amount in base units (decimal string).
    /// Mirrors `SubscriptionRequest.amount` so the activation builder
    /// can fill `SubscribeData.expected_amount` + `TransferData.amount`
    /// without re-parsing the parent request.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub amount: Option<String>,
    /// Base58 address of the subscriptions program deployment.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub subscription_program: Option<String>,
    /// Solana network slug — `mainnet`, `devnet`, `testnet`,
    /// `localnet`. Servers MAY accept `mainnet-beta` as a legacy alias
    /// of `mainnet` but MUST emit `mainnet`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub network: Option<String>,
    /// When `true`, the server pays activation transaction fees.
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub fee_payer: bool,
    /// Base58 of the fee-payer pubkey. REQUIRED when `fee_payer` is
    /// true.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fee_payer_key: Option<String>,
    /// Pre-fetched recent blockhash. When set, the client skips its own
    /// `getLatestBlockhash` RPC call.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recent_blockhash: Option<String>,
    /// On-chain `plan_id` (u64) the program reads from `SubscribeData`.
    /// The string `plan_id` above is the PDA derived from this number.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plan_id_numeric: Option<u64>,
    /// Plan PDA's bump seed.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plan_bump: Option<u8>,
    /// Plan's `period_hours` — `period_count * 24` for day, `* 168` for
    /// week.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_period_hours: Option<u64>,
    /// Plan's `created_at` unix timestamp (set on-chain by the program
    /// at Plan creation, read back into the YAML by `pay server start`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_created_at: Option<i64>,
}

impl SubscriptionMethodDetails {
    /// Decode from the parsed `methodDetails` JSON value.
    pub fn from_json(value: &serde_json::Value) -> Result<Self, Error> {
        serde_json::from_value(value.clone())
            .map_err(|e| Error::Other(format!("Invalid methodDetails: {e}")))
    }

    /// Validate the spec's REQUIRED fields are non-empty. The struct
    /// can be deserialised with missing required fields (serde will
    /// supply defaults), so callers that need a settle-able activation
    /// must run this check.
    pub fn validate(&self) -> Result<(), Error> {
        if self.plan_address.is_empty() {
            return Err(Error::Other("methodDetails.planAddress is required".into()));
        }
        if self.mint.is_empty() {
            return Err(Error::Other("methodDetails.mint is required".into()));
        }
        if self.token_program.is_empty() {
            return Err(Error::Other(
                "methodDetails.tokenProgram is required".into(),
            ));
        }
        if self.puller.is_empty() {
            return Err(Error::Other("methodDetails.puller is required".into()));
        }
        if self
            .subscription_program
            .as_deref()
            .is_none_or(str::is_empty)
        {
            return Err(Error::Other(
                "methodDetails.subscriptionProgram is required".into(),
            ));
        }
        if self.fee_payer && self.fee_payer_key.is_none() {
            return Err(Error::Other(
                "methodDetails.feePayerKey is required when feePayer is true".into(),
            ));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ed25519_dalek::{Signer, SigningKey};

    #[test]
    fn subscription_authentication_uses_canonical_message_and_binds_delegation() {
        let signing_key = SigningKey::from_bytes(&[7; 32]);
        let payer = bs58::encode(signing_key.verifying_key().as_bytes()).into_string();
        let delegation = "8tWbqLkUJoYy7zXc5h2EvCRoaQEv2xnQjUuYhc3rzCgT";
        let mut authentication = SubscriptionAuthentication {
            kind: SubscriptionAuthenticationType::Proof,
            challenge_id: "challenge-1".into(),
            payer: payer.clone(),
            signature: String::new(),
        };
        let message = authentication.message_bytes(delegation).unwrap();
        assert_eq!(
            String::from_utf8(message.clone()).unwrap(),
            format!(
                "{{\"domain\":\"mpp-subscription-auth-v1\",\"payer\":\"{payer}\",\"subscriptionChallengeId\":\"challenge-1\",\"subscriptionDelegation\":\"{delegation}\"}}"
            )
        );
        authentication.signature =
            bs58::encode(signing_key.sign(&message).to_bytes()).into_string();
        assert!(authentication.verify(delegation).unwrap());
        assert!(!authentication
            .verify("9xAXssX9j7vuK99c7cFwqbixzL3bFrzPy9PUhCtDPAYJ")
            .unwrap());
    }

    #[test]
    fn day_period_maps_to_hours() {
        assert_eq!(SubscriptionPeriodUnit::Day.to_period_hours(1).unwrap(), 24);
        assert_eq!(
            SubscriptionPeriodUnit::Day.to_period_hours(30).unwrap(),
            720
        );
        assert_eq!(
            SubscriptionPeriodUnit::Day.to_period_hours(365).unwrap(),
            8760
        );
    }

    #[test]
    fn week_period_maps_to_hours() {
        assert_eq!(
            SubscriptionPeriodUnit::Week.to_period_hours(1).unwrap(),
            168
        );
        assert_eq!(
            SubscriptionPeriodUnit::Week.to_period_hours(52).unwrap(),
            8736
        );
    }

    #[test]
    fn period_count_out_of_range_errors() {
        assert!(SubscriptionPeriodUnit::Day.to_period_hours(366).is_err());
        assert!(SubscriptionPeriodUnit::Week.to_period_hours(53).is_err());
        assert!(SubscriptionPeriodUnit::Day.to_period_hours(0).is_err());
    }

    #[test]
    fn request_serializes_camel_case() {
        let req = SubscriptionRequest {
            amount: "10000000".into(),
            currency: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v".into(),
            period_unit: SubscriptionPeriodUnit::Day,
            period_count: "30".into(),
            recipient: "9xAXssX9j7vuK99c7cFwqbixzL3bFrzPy9PUhCtDPAYJ".into(),
            subscription_expires: Some("2026-07-14T12:00:00Z".into()),
            external_id: Some("order-42".into()),
            ..Default::default()
        };
        let json = serde_json::to_string(&req).unwrap();
        assert!(json.contains("\"periodUnit\":\"day\""));
        assert!(json.contains("\"periodCount\":\"30\""));
        assert!(json.contains("\"subscriptionExpires\":\"2026-07-14T12:00:00Z\""));
        assert!(json.contains("\"externalId\":\"order-42\""));
    }

    #[test]
    fn rejects_unknown_period_unit_via_deserialization() {
        let json = r#"{"amount":"1","currency":"X","periodUnit":"month","periodCount":"1","recipient":"R"}"#;
        let parsed: Result<SubscriptionRequest, _> = serde_json::from_str(json);
        assert!(
            parsed.is_err(),
            "month must be rejected at the schema layer"
        );
    }

    #[test]
    fn period_hours_validates_range() {
        let req = SubscriptionRequest {
            period_unit: SubscriptionPeriodUnit::Day,
            period_count: "30".into(),
            ..Default::default()
        };
        assert_eq!(req.period_hours().unwrap(), 720);

        let too_big = SubscriptionRequest {
            period_unit: SubscriptionPeriodUnit::Day,
            period_count: "400".into(),
            ..Default::default()
        };
        assert!(too_big.period_hours().is_err());
    }

    #[test]
    fn activate_payload_pull_mode_roundtrip() {
        let action = SubscriptionAction::Activate(ActivatePayload {
            payload_type: "transaction".into(),
            transaction: Some("AQAAAA==".into()),
            signature: None,
            authentication: None,
        });
        let json = serde_json::to_string(&action).unwrap();
        assert!(json.contains("\"action\":\"activate\""));
        assert!(json.contains("\"type\":\"transaction\""));
        assert!(!json.contains("\"signature\""));
        let _back: SubscriptionAction = serde_json::from_str(&json).unwrap();
    }

    #[test]
    fn activate_payload_push_mode_roundtrip() {
        let action = SubscriptionAction::Activate(ActivatePayload {
            payload_type: "signature".into(),
            transaction: None,
            signature: Some("5J8KKKKK".into()),
            authentication: None,
        });
        let json = serde_json::to_string(&action).unwrap();
        assert!(json.contains("\"type\":\"signature\""));
        assert!(json.contains("\"signature\":\"5J8KKKKK\""));
        let _back: SubscriptionAction = serde_json::from_str(&json).unwrap();
    }

    #[test]
    fn parse_amount_and_period_count() {
        let req = SubscriptionRequest {
            amount: "10000000".into(),
            period_count: "30".into(),
            ..Default::default()
        };
        assert_eq!(req.parse_amount().unwrap(), 10_000_000);
        assert_eq!(req.parse_period_count().unwrap(), 30);

        let bad_amount = SubscriptionRequest {
            amount: "not-a-number".into(),
            period_count: "30".into(),
            ..Default::default()
        };
        assert!(bad_amount.parse_amount().is_err());

        let bad_count = SubscriptionRequest {
            amount: "1".into(),
            period_count: "abc".into(),
            ..Default::default()
        };
        assert!(bad_count.parse_period_count().is_err());
    }

    #[test]
    fn period_hours_bound_check() {
        // day*365 = 8760 — at the upper bound, accepted.
        let at_bound = SubscriptionRequest {
            period_unit: SubscriptionPeriodUnit::Day,
            period_count: "365".into(),
            ..Default::default()
        };
        assert_eq!(at_bound.period_hours().unwrap(), 8760);

        // Bad period_count surfaces through period_hours().
        let bad = SubscriptionRequest {
            period_unit: SubscriptionPeriodUnit::Day,
            period_count: "not-a-number".into(),
            ..Default::default()
        };
        assert!(bad.period_hours().is_err());
    }

    #[test]
    fn activate_payload_default() {
        let p = ActivatePayload::default();
        assert!(p.transaction.is_none());
        assert!(p.signature.is_none());
        assert_eq!(p.payload_type, "");
    }

    #[test]
    fn period_unit_default_is_day() {
        let unit = SubscriptionPeriodUnit::default();
        assert_eq!(unit, SubscriptionPeriodUnit::Day);
    }

    #[test]
    fn receipt_extensions_default() {
        let ext = SubscriptionReceiptExtensions::default();
        assert_eq!(ext.subscription_id, "");
        assert_eq!(ext.subscription_delegation, "");
        assert!(ext.expires_at.is_none());
    }

    #[test]
    fn receipt_extensions_serialize() {
        let ext = SubscriptionReceiptExtensions {
            subscription_id: "BXQGmO5VwTrl5RfFr6Y8XQZ4nPj9QqMOiKkRn3pZ4ZE".into(),
            subscription_delegation: "De1egation".into(),
            period_index: 0,
            period_start: "2026-01-15T12:03:10Z".into(),
            period_end: "2026-02-14T12:03:10Z".into(),
            expires_at: Some("2026-07-14T12:00:00Z".into()),
        };
        let json = serde_json::to_string(&ext).unwrap();
        assert!(json.contains("\"subscriptionId\""));
        assert!(json.contains("\"subscriptionDelegation\""));
        assert!(json.contains("\"periodIndex\":0"));
        assert!(json.contains("\"periodStart\""));
        assert!(json.contains("\"periodEnd\""));
        assert!(json.contains("\"expiresAt\""));
    }
}

File: python/src/solana_pay_kit/protocols/mpp/client/subscription.py
"""Client-side subscription activation and access credentials.

Activation builds one subscriber-signed transaction that creates the
``SubscriptionDelegation`` and collects the first period:
``[compute budget, init authority?, subscribe, transfer_subscription, memo?]``.
The on-chain ``Plan`` is read and checked against the challenge before
anything is signed, so a server cannot steer the subscriber into a plan whose
terms, destination or puller differ from what the challenge advertises.

The ``SubscriptionAuthority`` is initialized in the same transaction only when
it is missing, with the ``UNKNOWN_INIT_ID`` sentinel so ``subscribe`` binds the
authority created in that slot. No associated-token-account instruction is
emitted: the subscriber must already hold the mint.
"""

from __future__ import annotations

import base64
import time
from dataclasses import dataclass
from typing import Any

from solders.hash import Hash  # type: ignore[import-untyped]
from solders.instruction import Instruction  # type: ignore[import-untyped]
from solders.pubkey import Pubkey  # type: ignore[import-untyped]

from solana_pay_kit._paycore.solana import COMPUTE_BUDGET_PROGRAM, MEMO_PROGRAM
from solana_pay_kit._paycore.transaction import build_partially_signed_v0_transaction
from solana_pay_kit.protocols.mpp._subscriptions import (
    SUBSCRIPTIONS_PROGRAM_ID,
    UNKNOWN_INIT_ID,
    PlanView,
    authority_init_id,
    build_init_subscription_authority_ix,
    build_subscribe_ix,
    build_transfer_subscription_ix,
    decode_plan,
    find_subscription_authority_pda,
    find_subscription_pda,
    plan_problems,
    sign_message,
    signer_pubkey,
)
from solana_pay_kit.protocols.mpp.core.expires import parse_rfc3339
from solana_pay_kit.protocols.mpp.core.types import ChallengeEcho, PaymentChallenge, PaymentCredential
from solana_pay_kit.protocols.mpp.intents.subscription import (
    AccessPayload,
    ActivatePayload,
    PeriodUnit,
    SubscriptionAuthentication,
    SubscriptionMethodDetails,
    SubscriptionRequest,
    sign_subscription_authentication,
)

__all__ = [
    "SubscriptionActivation",
    "build_subscription_access_credential",
    "build_subscription_activation",
]

_MAX_MEMO_BYTES = 566


@dataclass(frozen=True)
class SubscriptionActivation:
    """An activation credential plus the bearer proof to keep for later access."""

    credential: PaymentCredential
    authentication: SubscriptionAuthentication
    subscription_delegation: str


def _same_on_the_wire(advertised: object, actual: object) -> bool:
    """Compare a methodDetails field with the chain at the precision canonical JSON keeps.

    RFC 8785 renders every JSON number as an ECMAScript double (ECMA-262 7.1.12.1),
    so a ``planIdNumeric`` above 2**53 reaches the client rounded from any server
    that canonicalizes the request. Comparing the digits would refuse those plans;
    comparing the doubles still pins about 16 significant digits, and the
    activation itself is built from the plan account, never from this number.
    """
    if isinstance(advertised, int) and isinstance(actual, int):
        return float(advertised) == float(actual)
    return advertised == actual


def _extension_problems(details: SubscriptionMethodDetails, request: SubscriptionRequest, plan: PlanView) -> list[str]:
    """Cross-check the optional Rust plan extensions against the chain; absent fields are skipped."""
    checks: list[tuple[str, object, object]] = [
        ("merchant", details.merchant, str(plan.owner)),
        ("recipient", details.recipient, request.recipient),
        ("amount", details.amount, request.amount),
        ("planIdNumeric", details.plan_id_numeric, plan.plan_id),
        ("planBump", details.plan_bump, plan.bump),
        ("expectedPeriodHours", details.expected_period_hours, plan.period_hours),
        ("expectedCreatedAt", details.expected_created_at, plan.created_at),
    ]
    return [
        f"methodDetails.{name} is {advertised!r} but the on-chain plan has {actual!r}"
        for name, advertised, actual in checks
        if advertised not in ("", None) and not _same_on_the_wire(advertised, actual)
    ]


async def build_subscription_activation(
    signer: Any,
    rpc: Any,
    challenge: PaymentChallenge,
    *,
    subscription_program: str = SUBSCRIPTIONS_PROGRAM_ID,
    compute_unit_limit: int = 400_000,
    compute_unit_price: int = 1,
    max_amount_base_units: int | None = None,
    expected_currency: str | None = None,
    expected_recipient: str | None = None,
    expected_period_unit: PeriodUnit | None = None,
    expected_period_count: int | None = None,
) -> SubscriptionActivation:
    """Check the challenge against the on-chain plan, then build and sign the activation credential.

    ``signer`` is a solana_pay_kit signer or a solders ``Keypair``; ``rpc`` is a
    :class:`~solana_pay_kit._paycore.rpc.SolanaRpc`-compatible client
    (``get_account_info`` returning ``(data, owner)`` and ``get_latest_blockhash``).
    ``subscription_program`` is the deployment the client trusts; a challenge
    naming another program is refused before any RPC call. The ``max_amount``
    and ``expected_*`` guards are the spec's amount and period verification:
    each one that is set must hold, and they are checked before any RPC call.
    A challenge whose ``subscriptionExpires`` has passed is always refused.
    """
    if challenge.method != "solana" or challenge.intent != "subscription":
        raise ValueError("challenge is not a solana subscription challenge")
    if challenge.is_expired():
        raise ValueError(f"challenge expired at {challenge.expires}")
    request = SubscriptionRequest.from_dict(challenge.decode_request())
    details = request.method_details
    if details.subscription_program != subscription_program:
        raise ValueError(f"challenge names subscriptions program {details.subscription_program}, not the trusted one")
    if len(request.external_id.encode("utf-8")) > _MAX_MEMO_BYTES:
        raise ValueError(f"externalId cannot exceed {_MAX_MEMO_BYTES} bytes")
    guards: list[tuple[str, bool]] = [
        (
            "amount exceeds max_amount_base_units",
            max_amount_base_units is not None and int(request.amount) > max_amount_base_units,
        ),
        ("currency", expected_currency is not None and request.currency != expected_currency),
        ("recipient", expected_recipient is not None and request.recipient != expected_recipient),
        ("periodUnit", expected_period_unit is not None and request.period_unit != expected_period_unit),
        ("periodCount", expected_period_count is not None and int(request.period_count) != expected_period_count),
    ]
    failed = [name for name, broken in guards if broken]
    if failed:
        raise ValueError("refusing to sign: challenge " + ", ".join(failed) + " does not match expectations")
    if request.subscription_expires and parse_rfc3339(request.subscription_expires).timestamp() <= time.time():
        raise ValueError(f"refusing to sign: subscriptionExpires {request.subscription_expires} has passed")

    program = Pubkey.from_string(subscription_program)
    subscriber = signer_pubkey(signer)
    plan_address = Pubkey.from_string(details.plan_address)
    account = await rpc.get_account_info(details.plan_address)
    if account is None:
        raise ValueError(f"plan {details.plan_address} does not exist")
    plan = decode_plan(account[0], account[1], subscription_program, plan_address)
    problems = plan_problems(
        plan,
        mint=details.mint,
        amount=int(request.amount),
        period_hours=request.period_hours(),
        recipient=request.recipient,
        puller=details.puller,
        now=int(time.time()),
    ) + _extension_problems(details, request, plan)
    if problems:
        raise ValueError(f"refusing to sign plan {details.plan_address}: " + "; ".join(problems))

    mint = Pubkey.from_string(details.mint)
    token_program = Pubkey.from_string(details.token_program)
    compute_budget = Pubkey.from_string(COMPUTE_BUDGET_PROGRAM)
    instructions = [
        Instruction(compute_budget, bytes([2]) + compute_unit_limit.to_bytes(4, "little"), []),
        Instruction(compute_budget, bytes([3]) + compute_unit_price.to_bytes(8, "little"), []),
    ]
    authority = await rpc.get_account_info(str(find_subscription_authority_pda(subscriber, mint, program)))
    init_id = authority_init_id(authority, subscription_program)
    if init_id is None:
        instructions.append(
            build_init_subscription_authority_ix(
                program=program, subscriber=subscriber, mint=mint, token_program=token_program
            )
        )
        init_id = UNKNOWN_INIT_ID

    fee_payer = Pubkey.from_string(details.fee_payer_key) if details.fee_payer else subscriber
    instructions.append(
        build_subscribe_ix(
            program=program,
            subscriber=subscriber,
            plan=plan,
            init_id=init_id,
            payer=fee_payer if details.fee_payer else None,
        )
    )
    instructions.append(
        build_transfer_subscription_ix(
            program=program,
            subscriber=subscriber,
            plan=plan,
            recipient=Pubkey.from_string(request.recipient),
            puller=Pubkey.from_string(details.puller),
            token_program=token_program,
            amount=plan.amount,
        )
    )
    if request.external_id:
        instructions.append(Instruction(Pubkey.from_string(MEMO_PROGRAM), request.external_id.encode("utf-8"), []))

    blockhash = details.recent_blockhash or (await rpc.get_latest_blockhash()).value.blockhash
    wire = build_partially_signed_v0_transaction(
        instructions,
        fee_payer,
        Hash.from_string(str(blockhash)),
        subscriber,
        lambda message: sign_message(signer, message),
    )
    delegation = str(find_subscription_pda(plan_address, subscriber, program))
    authentication = sign_subscription_authentication(challenge.id, delegation, signer)
    payload = ActivatePayload(transaction=base64.b64encode(wire).decode("ascii"), authentication=authentication)
    credential = PaymentCredential(
        challenge=challenge.to_echo(),
        payload=payload.to_dict(),
        source=f"did:pkh:solana:{details.network}:{subscriber}",
    )
    return SubscriptionActivation(
        credential=credential, authentication=authentication, subscription_delegation=delegation
    )


def build_subscription_access_credential(
    challenge: ChallengeEcho, subscription_delegation: str, authentication: SubscriptionAuthentication
) -> PaymentCredential:
    """Build the ``type="proof"`` access credential that echoes the activation challenge."""
    payload = AccessPayload(subscription_delegation=subscription_delegation, authentication=authentication)
    return PaymentCredential(challenge=challenge, payload=payload.to_dict())
